README.md
# Simple MCP4922 SPI DAC library for Teensy 3.2 boards.

## Simple MCP4922

This library provides a simple and fast way to set up a MCP4922 DAC on a Teensy 3.2 board. The code is simple and short, it was designed to fulfill an specific need for a project, disabling some of the hardware pins used by the SPI interface. But it ended up being *kind of* redundant since (*re*)setting the pin does the job in a similar way. 

This library is based on [Bob's Tillart MCP_DAC library](https://github.com/RobTillaart/MCP_DAC) while doing some of the operations in a slightly different way.

## Functionalities

### Constructor 

- **Simple_MCP4922()** The constuctor intializes some of the values to the logical default ones such as gain=1, alternatePins=false etc.
- **init()** Function called in the constructor to intialize the values.
- **begin(uint8_t cs, bool alternatePins)** Recieves a pin to use as CS or Chip Select since the SPI library doesn't take care of it. Also a boolean value that controls the use of some of the altenate pins designated in the interface.

### Settings

- **setBuffer(bool buffered)** *True* enables buffered outputs while *False* disables them.
- **setGain(uint8_t gain)** The attribute *_gain* should be either 1 or 2, usually I would have had 0 or 1 for gain 1 and 2 respectively, but the DAC sets gain to 1 when sent a high level bit, so I thought it was more convenient this way. 
- **end()** can be used to call SPI.end() to disable the interface.

### Disabling certain pins of the interface

A SPI hardware interface tipically has 4 pins: MISO for data input, MOSI for data output, CSK for the clock and CS for selecting the peripheral device.

For example, seting MISO pin as pinMode(CS, OUTPUT) will give the GPIO the control of the pin even tho the SPI interface was using it. Athough this *works*, I included dedicated functions to change the pin control "completely". A multiplexor sets which system controls each pin, so its possible to write to the register in order to modify which system can use it.

```cpp
// PORTx_PCRn, for pin 12 (default MOSI):
PORTC_PCR7 = (PORTC_PCR7 & 0xFFFFF8FF) | (001 << 8);
```

Bits 8-10 in the register control the multiplexor as shown in the [manual](https://www.pjrc.com/teensy/K20P64M72SF1RM.pdf) for Tensy 3.2 on page 228.

- **freeMOSI()** Gives GPIO the control of pin 11 or pin 7 if alternatePins is *True*.
- **freeMISO()** Gives GPIO the control of pin 12 or pin 8 if alternatePins is *True*.
- **freeCSK()** Gives GPIO the control of pin 13 or pin 14 if alternatePins is *True*.
- Freeing CS is not necessary since its not controlled by the library.

### Ouputting data

- **analogWrite(uint16_t data, uint8_t channel)** Channel can be either 0 (OutputA) or 1 (OutputB).
- **analogWriteFastCH1(uint16_t data)** and **analogWriteFastCH2(uint16_t data)** ignore buffer and gain settings and use the default values (unbuffered and G=1).

## Setting up the DAC

**SHDN** pin should be connected to high level since it enables the output. If you have spare pins a microcontroller pin can be used to switch between *enabled* and *disabled* in order to save power.

The rest of the pins are listed in the datasheet, a full list of connections for basic operation is provided below.

| DAC   | PIN | Connection                |
|-------|-----|---------------------------|
| VDD   | 1   | VDD                       |
| NC    | 2   | x                         |
| !CS   | 3   | CS: Teensy3.2 Pin 10      |
| SCK   | 4   | SCK: Teensy 3.2 Pin 13    |
| SDI   | 5   | SDI: Teensy 3.2 Pin 11    |
| NC    | 6   | x                         |
| NC    | 7   | x                         |
| !LDAC | 8   | GND                       |
| !SHDN | 9   | VDD                       |
| VoutB | 10  | Channel 2 Ouput           |
| VrefB | 11  | Channel 2 Reference (VDD) |
| VSS   | 12  | GND                       |
| VrefA | 13  | Channel 1 Reference (VDD) |
| VrefB | 14  | Channel 1 Output          |

The code below shows *Simple_MCP4922_example.ino*. It's written to output values on channel A (or 1), connections to the DAC should be as shown in the table above.

```cpp
#include "Simple_MCP4922.h"

#define T_PIN_CS 10

uint16_t data_array[6] = {0, 819, 1638, 2457, 3276, 4095};
uint8_t i = 0;

// Create a Simple_MCP4922 object.
Simple_MCP4922 DAC;

void setup() 
{    
  // Start the SPI hardware controller.
  // The Simple_MCP4922 library takes care of importing
  // the SPI and setting up the chip select pin.
  DAC.begin(T_PIN_CS, false);
}

void loop() 
{

  // Writes to channel 1
  DAC.analogWrite(data_array[i], 0);
  if(i < 5) i++;
  else i = 0;

  delay(600); //msas
  
}
```

## Future

Possibly adding support for different Teensy boards.


changelog.txt
CHANGE LOG
*****************************************

Date
*********************
Description: 

  Added:
    -
    -
  
  Removed:
    -
    -

  Changed:
    -
    -

  Notes:
    -
    -

Date
*********************
Description: 

  Added:
    -
    -
  
  Removed:
    -
    -

  Changed:
    -
    -

  Notes:
    -
    -



Simple_MCP4922.h
//#pragma once

#ifndef __SIMPLE_MCP4922_H__
#define __SIMPLE_MCP4922_H__

//#if defined(__arm__) && defined(TEENSYDUINO) && defined(KINETISK)

#include "Arduino.h"
#include <SPI.h>

#define MCP4922_MAX 4095

class Simple_MCP4922
{ 
public:
  Simple_MCP4922();

  void begin(uint8_t CS, bool altenratePins);

  void init();

  //Altenernate CS pin considered as pin 9 only
  void setAlternatePins(bool MOSI, bool MISO, bool SCK, bool CS);
  
  bool analogWrite(uint16_t data, uint8_t channel);

  void analogWriteFastCH1(uint16_t data);

  void analogWriteFastCH2(uint16_t data);
  
  // Data outputs can be buffered (true) or non-buffered (false)
  void setBuffer(bool buffered);

  //  Gain can be 1 or 2,  (maximum output voltage can never exceed VCC level)
  void setGain(uint8_t gain);

  void freeMOSI();

  void freeMISO();

  void freeSCK();

  //SPI library doesn't control the CS pin.
  //void freeCS();

  void end();

protected:

  uint8_t _pinCS; //SPI library doesn't control the CS pin.

  bool    _buffered;
  uint8_t _gain;

  bool    _alternatePins;
  bool    _altMISO;
  bool    _altMOSI;
  bool    _altSCK;
  bool    _altCS;

  void    transfer(uint16_t data);

};

//#endif
#endif

Simple_MCP4922.cpp
#include "Simple_MCP4922.h"

Simple_MCP4922::Simple_MCP4922()
{
  init();
}

void Simple_MCP4922::init()
{
  _pinCS         = 10;

  _buffered      = false;
  _gain          = 1;

  _alternatePins = false;
  _altMISO       = false;
  _altMOSI       = false;
  _altSCK        = false;
  _altCS         = false;
}

void Simple_MCP4922::setAlternatePins(bool MOSI, bool MISO, bool SCK, bool CS)
{
  _altMOSI = MOSI;
  _altMISO = MISO;
  _altSCK  = SCK;
  //_altCS   = CS;
}

void Simple_MCP4922::begin(uint8_t CS, bool alternatePins)
{
  _pinCS         =  CS;
  _alternatePins = alternatePins;

  //Set CS to NOT-SELECTED (usually CS pin is active low)
  pinMode(_pinCS, OUTPUT);
  digitalWrite(_pinCS, HIGH);

  if(_alternatePins)
  {
    #if defined(__MK20DX256__)
      //SPI.end(); //Maybe delete for faster operation
      if(_altMOSI)  SPI.setMOSI(7);
      if(_altMISO)  SPI.setMISO(8);
      if(_altSCK)   SPI.setSCK(14);
      //if(_altCS)    SPI.setCS(9);
      SPI.begin();
    #endif
  }
  else
  {
    //According to https://forum.pjrc.com/threads/49562-Audio-functionality-and-SPI-communication-again!-(Teensy-3-2)
    //SPI library doesnt save alternate pin changes so no there is no need to reset them.
    //SPI.end();

    SPI.begin();
  }

  //Apparently this type of settings is deprecated https://arduino.stackexchange.com/questions/79728/when-is-spi-begintransaction-required
  //SPI.setClockDivider(SPI_CLOCK_DIV2); //24MHz (MPC4922 rated for 20MHz)
}

bool Simple_MCP4922::analogWrite(uint16_t data, uint8_t channel)
{

  //| CH | Buff | G  | SH | D11 | D10 | D9 | D8 | D7 | D6 | D5 | D4 | D3 | D2 | D1 | D0 |
  //| 15 | 14   | 13 | 12 | 11  | 10  | 9  | 8  | 7  | 6  | 5  | 4  | 3  | 2  | 1  | 0  |

  //data &= 0x0FFF;         //Cleans the top bits 
  data |= (channel  << 15);
  data |= (1        << 12); //Sets SHDN to 1 (enable output).
  
  if(_buffered)        data |= (1  << 14);
  else                 data |= (0  << 14);

  if(_gain == 1)       data |= (1 << 13);
  else if (_gain == 2) data |= (0 << 13);

  transfer(data);

  return true;
}

void Simple_MCP4922::analogWriteFastCH1(uint16_t data)
{
  // 0x3 = 0011 (CH=0, BUFF=0, GA=1, SHDN=1)
  transfer(data | 0x3000); 
}

void Simple_MCP4922::analogWriteFastCH2(uint16_t data)
{
  // 0x3 = 1011 (CH=1, BUFF=0, GA=1, SHDN=1)
  transfer(data | 0xB000);
}

void Simple_MCP4922::transfer(uint16_t data)
{
  //SPI.transfer sends a byte of data so we need to split it
  uint8_t data_high = data >> 8;   //Gets the 8 MSBs
  uint8_t data_low  = data & 0xFF; //Gets the 8 LSBs

  digitalWrite(_pinCS, LOW); //Select chip 

  SPI.beginTransaction(SPISettings(20000000, MSBFIRST, SPI_MODE0));
  SPI.transfer(data_high);
  SPI.transfer(data_low);
  SPI.endTransaction();

  digitalWrite(_pinCS, HIGH);
}

void Simple_MCP4922::setBuffer(bool buffered)
{
  _buffered = buffered;
}

void Simple_MCP4922::setGain(uint8_t gain)
{
  _gain =  gain;
}

void Simple_MCP4922::freeMOSI()
{
  #if defined(__MK20DX256__)
  if(_alternatePins)
  {
    //PIN 7
    PORTD_PCR2 = (PORTD_PCR2 & 0xFFFFF8FF) | (001 << 8);
  }
  else
  {
    //PIN 11
    PORTC_PCR6 = (PORTC_PCR6 & 0xFFFFF8FF) | (001 << 8);
  }
  #endif
}

void Simple_MCP4922::freeMISO(){
  #if defined(__MK20DX256__)
  if(_alternatePins)
  {
    //PIN 8
    PORTD_PCR3 = (PORTD_PCR3 & 0xFFFFF8FF) | (001 << 8);
  }
  else
  {
    //PIN 12
    PORTC_PCR7 = (PORTC_PCR7 & 0xFFFFF8FF) | (001 << 8);
  }
  #endif
}

void Simple_MCP4922::freeSCK(){
  #if defined(__MK20DX256__)
  if(_alternatePins)
  {
    //PIN 14
    PORTD_PCR1 = (PORTD_PCR1 & 0xFFFFF8FF) | (001 << 8);
  }
  else
  {
    //PIN 13
    PORTC_PCR5 = (PORTC_PCR5 & 0xFFFFF8FF) | (001 << 8);
  }
  #endif
}

/*
void freeCS(){
  #if defined(__MK20DX256__)
  if(_alternatePins)
  {
    //PIN 9
    PORTC_PCR3 = (PORTC_PCR3 & 0xFFFFF8FF) | (001 << 8);
  }
  else
  {
    //PIN 10
    PORTC_PCR4 = (PORTC_PCR4 & 0xFFFFF8FF) | (001 << 8);
  }
  #endif
}
*/

void Simple_MCP4922::end()
{
  SPI.end();
}